Thorlabs APT Inertial Actuator

trappitsch opened the pull request:
Added Thorlabs APT Inertial Actuator device for control of TIM101 and KIM101 piezo motors. I have seen that there were many many discussions on the APT going on... opening this PR to discuss some current issues, look for ideas, and check on what you remember from these former discussions that could come in handy:

**Device integration:**
- The TIM and KIM controllers deviate quite a bit in commands from the standard Thorlabs APT controllers, so I decided for a new class. Inheriting the APT classes, but not all parent functions work. Added some catchers for when not, but it's not that pretty.
- TIM and KIM controllers also deviate... Only have a KIM101 currently available, that's the new one though. Catching when commands only apply to KIM*01 or KIM101.
- Ideas on how to improve this?

**Testing:**
I noticed that so far only the initialization was tested for, but none of the commands. So I couldn't just copy. The initialization makes it difficult to use the `with_expected_protocol()` routine to test a command. My fix / workaround for now:

- Write an `@pytest.fixture` that simply returns the initialization `stdin` and `stdout` for the KIM101 controller.
- Use this fixture to initialize it in every test

So far I have only written one test, mostly as a proof of principle that this idea works. However, is it the greatest idea? Fairly certain it's not. I wanted to see if you have more ideas. Once we have a nice solution, I'll write more tests to get the APT covered. The good news, I currently also have a KDC101 (Brushed DC Motor controller) with a PRM1-Z8 around for testing other routines and writing tests. More than happy to check the routines and write tests for the existing `APTMotorController` as well and implement them.

@scasagrande Hope you enjoyed your vacation!

trappitsch commented on instruments/tests/test_thorlabs/test_thorlabs_apt.py:
```diff
@@ -79,3 +79,72 @@ def test_apt_hw_info():
+@pytest.fixture
+def init_kim101():
+    """Return the send, receive value to initialize a KIM101 unit."""
+    stdin = ThorLabsPacket(
+        message_id=ThorLabsCommands.HW_REQ_INFO,
+        param1=0x00, param2=0x00,
+        dest=0x50,
+        source=0x01,
+        data=None
+    ).pack()
+    stdout = ThorLabsPacket(
+        message_id=ThorLabsCommands.HW_GET_INFO,
+        dest=0x01,
+        source=0x50,
+        data=hw_info_data.pack(
+            # Serial number
+            b'\x01\x02\x03\x04',
+            # Model number
+            "KIM101".encode('ascii'),
+            # HW type
+            16,
+            # FW version,
+            0xa1, 0xa2, 0xa3,
+            # Notes
+            "abcdefg".encode('ascii'),
+            # HW version
+            42,
+            # Mod state
+            43,
+            # Number of channels
+            4
+        )
+    ).pack()
```
These are just the same parameters as in the existing test, but changed for a KIM101 adapter.
 - Model Number
 - Number of Channels

trappitsch commented on instruments/thorlabs/thorlabsapt.py:
```diff
@@ -52,7 +54,17 @@ def enabled(self):
+
+            :raises TypeError: If controller is not supported
             """
+            if self._apt.model_number[0:3] == 'KIM':
+                raise TypeError("For KIM controllers, use the "
+                                "`enabled_single` function to enable "
+                                "one axis. For KIM101 controllers, "
+                                "multiple axes can be enabled using "
+                                "the `enabled_multi` function from the "
+                                "controller level.")
```
According to the APT manual, the KIM101 should react to this command. However, mine does not whatsoever. In the new messaging system that is implemented, they have now commands to enable axes. I did not want to re-implement `enabled` because it might work with a KIM001 or a TIM101 controller.

scasagrande commented:
You are on a roll!

You're right about the Thorlabs equipment being a pain. I have distinct memories from back in 2011-2013 of trying to get the APT modules to work, trying to get more help and documentation from Thorlabs, etc. 

You should be able to use the stdin and stdout that you've defined with your pytest fixture in the `with_expected_protocol` framework. It might not be the best, but then again nothing these modules do is.

trappitsch commented:
Since the pandemic keeps me from getting measurements in lab done, at least I can improve the handling of everything. Turns out there was a lot to catch up on in terms of software.

For Thorlabs APT I'm going to start implementing the tests then and incorporate them for the two motor cubes that I have here with me now. That will hopefully help in the future with further development as well. The class as it is right now seems to work fairly well for both cubes. I will only implement tests that I can test directly on the hardware I have here and note with what it was tested, to keep a record. Might not give 100% coverage, but it's probably the safer route. I'll put some more effort into writing docstring examples as well to help users (and myself...) in the future.

The PDF with the comms protocols on the APT devices is a bit newer now and available from the thorlabs website. It's issue 25 now (the one in the repo is issue 9) I saw that the new one is structured a bit better to account for more and newer drivers, however, it still has many errors in it. What do you think, should we swap the pdf in the repo, add the new one, or just put a link to the thorlabs website and where to get it from and remove the existing pdf? 

I'll update the PR once I have the tests running and ready to go and rebase it properly too. Sorry, I somehow screwed up the the rebase for this one. More soon!

scasagrande commented:
Awesome!! :man_dancing: 

If you implement features without being able to manually validate them, can you include a note in their docstrings indicating such?

trappitsch commented:
Will do so!

trappitsch commented:
Finished up the test suite as discussed. There's currently no new routines that I implemented that were not tested with hardware.
Also went through the `APTMotorController` and fixed too bugs. Details are in the commit text.

Added some examples on how to talk to the motors and use different commands in the docstring. Hope those are appropriate and help in the future.

Pylint throws me nine depreciation warnings, all for invalid escape characters:

    .tox/py38/lib/python3.8/site-packages/vxi11/vxi11.py:135
      /home/reto/Documents/code/InstrumentKit/.tox/py38/lib/python3.8/site-packages/vxi11/vxi11.py:135: DeprecationWarning: invalid escape sequence \s
        '(::(?P<arg2>[^\s:]+(\[.+\])?))?(::(?P<suffix>INSTR))$',

    -- Docs: https://docs.pytest.org/en/latest/warnings.html

I didn't look further into this for today...

Let me know what you think about the APT comms manual and what to do with the old one that is in the repo. I'm gonna attach the latest version to this post to have a record.

[APT_Communications_Protocol.pdf](https://github.com/Galvant/InstrumentKit/files/4906168/APT_Communications_Protocol.pdf)


Oh, I rebased my commit again to have an real commit message this time. 

Happy weekend!

scasagrande commented:
First look seems that everything looks good to me!

Can we add these classes to the docs so they're auto generated? Just need to copy paste what you already see in the RST files.

trappitsch commented:
Added the new `APTInertialActuator` class to the docs and fixed some docstring formatting errors after building the docs locally.

scasagrande commented:
Looks great to me then. Anything else to add before we merge?

trappitsch commented:
Not that I can think of for the APT. Should run more stable and we got some more tests :) Thanks @scasagrande

scasagrande commented:
SHIP IT
